code/modules/lighting/lighting_atom.dm

```diff
@@ -41,25 +41,40 @@
 			light = new/datum/light_source(src, .)
 
 
-// Should always be used to change the opacity of an atom.
-// It notifies (potentially) affected light sources so they can update (if needed).
+/**
+  * Updates the atom's opacity value.
+  *
+  * This exists to act as a hook for associated behavior.
+  * It notifies (potentially) affected light sources so they can update (if needed).
+  */
 /atom/proc/set_opacity(new_opacity)
 	if (new_opacity == opacity)
 		return
-
+	SEND_SIGNAL(src, COMSIG_ATOM_SET_OPACITY, new_opacity)
+	. = opacity
 	opacity = new_opacity
-	var/turf/T = loc
-	if (!isturf(T))
+
+
+/atom/movable/set_opacity(new_opacity)
+	. = ..()
+	if(isnull(.) || !isturf(loc))
 		return
 
-	if (new_opacity == TRUE)
-		T.has_opaque_atom = TRUE
-		T.reconsider_lights()
+	var/turf/turf_loc = loc
+
+	if(opacity)
+		AddElement(/datum/element/light_blocking)
+		turf_loc.add_opacity_source(src)
```

Comment:
I feel like this should be handled by the light_blocking element itself. Same for calling remove_opacity_source() upon removal of the element, obviously.

Reply:
Sure